File: sdk/python/aragora/namespaces/usage.py
"""
Usage Namespace API

Provides methods for usage tracking and SME dashboard:
- Usage summary and breakdown
- ROI analysis
- Budget tracking
- Forecasting
- Benchmarks
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

BenchmarkType = Literal["sme", "enterprise", "tech_startup", "consulting"]
ExportFormat = Literal["csv", "json"]


class UsageAPI:
    """
    Synchronous Usage API for SME dashboard.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="your-key")
        >>> summary = client.usage.get_summary()
        >>> print(f"Total debates: {summary['total_debates']}")
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def get_summary(
        self,
        period: str = "month",
        organization_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Get unified usage metrics summary.

        Args:
            period: Time period (day, week, month, quarter, year)
            organization_id: Optional organization filter

        Returns:
            Usage summary with debates, tokens, costs, etc.
        """
        params: dict[str, Any] = {"period": period}
        if organization_id:
            params["organization_id"] = organization_id
        return self._client.request("GET", "/api/v1/usage/summary", params=params)

    def get_breakdown(
        self,
        group_by: str = "agent",
        period: str = "month",
    ) -> dict[str, Any]:
        """
        Get usage breakdown by dimension.

        Args:
            group_by: Dimension to group by (agent, model, day, debate_type)
            period: Time period

        Returns:
            Breakdown data by the specified dimension
        """
        return self._client.request(
            "GET",
            "/api/v1/usage/breakdown",
            params={"group_by": group_by, "period": period},
        )

    def get_roi(
        self,
        benchmark_type: BenchmarkType = "sme",
    ) -> dict[str, Any]:
        """
        Get ROI analysis with industry benchmarks.

        Args:
            benchmark_type: Benchmark to compare against

        Returns:
            ROI metrics with time saved, cost analysis, etc.
        """
        return self._client.request(
            "GET",
            "/api/v1/usage/roi",
            params={"benchmark_type": benchmark_type},
        )

    def get_budget_status(self) -> dict[str, Any]:
        """
        Get current budget utilization status.

        Returns:
            Budget status with limits, usage, and remaining
        """
        return self._client.request("GET", "/api/v1/usage/budget-status")

    def get_forecast(
        self,
        days: int = 30,
    ) -> dict[str, Any]:
        """
        Get usage forecast.

        Args:
            days: Number of days to forecast

        Returns:
            Projected usage and costs
        """
        return self._client.request(
            "GET",
            "/api/v1/usage/forecast",
            params={"days": days},
        )

    def get_benchmarks(
        self,
        benchmark_type: BenchmarkType = "sme",
    ) -> dict[str, Any]:
        """
        Get industry benchmarks for comparison.

        Args:
            benchmark_type: Industry benchmark type

        Returns:
            Benchmark data for comparison
        """
        return self._client.request(
            "GET",
            "/api/v1/usage/benchmarks",
            params={"type": benchmark_type},
        )

    def export(
        self,
        format: ExportFormat = "csv",
        period: str = "month",
    ) -> dict[str, Any]:
        """
        Export usage data.

        Args:
            format: Export format (csv or json)
            period: Time period to export

        Returns:
            Export data or download URL
        """
        return self._client.request(
            "GET",
            "/api/v1/usage/export",
            params={"format": format, "period": period},
        )


class AsyncUsageAPI:
    """
    Asynchronous Usage API for SME dashboard.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     summary = await client.usage.get_summary()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def get_summary(
        self,
        period: str = "month",
        organization_id: str | None = None,
    ) -> dict[str, Any]:
        """Get unified usage metrics summary."""
        params: dict[str, Any] = {"period": period}
        if organization_id:
            params["organization_id"] = organization_id
        return await self._client.request("GET", "/api/v1/usage/summary", params=params)

    async def get_breakdown(
        self,
        group_by: str = "agent",
        period: str = "month",
    ) -> dict[str, Any]:
        """Get usage breakdown by dimension."""
        return await self._client.request(
            "GET",
            "/api/v1/usage/breakdown",
            params={"group_by": group_by, "period": period},
        )

    async def get_roi(
        self,
        benchmark_type: BenchmarkType = "sme",
    ) -> dict[str, Any]:
        """Get ROI analysis with industry benchmarks."""
        return await self._client.request(
            "GET",
            "/api/v1/usage/roi",
            params={"benchmark_type": benchmark_type},
        )

    async def get_budget_status(self) -> dict[str, Any]:
        """Get current budget utilization status."""
        return await self._client.request("GET", "/api/v1/usage/budget-status")

    async def get_forecast(
        self,
        days: int = 30,
    ) -> dict[str, Any]:
        """Get usage forecast."""
        return await self._client.request(
            "GET",
            "/api/v1/usage/forecast",
            params={"days": days},
        )

    async def get_benchmarks(
        self,
        benchmark_type: BenchmarkType = "sme",
    ) -> dict[str, Any]:
        """Get industry benchmarks for comparison."""
        return await self._client.request(
            "GET",
            "/api/v1/usage/benchmarks",
            params={"type": benchmark_type},
        )

    async def export(
        self,
        format: ExportFormat = "csv",
        period: str = "month",
    ) -> dict[str, Any]:
        """Export usage data."""
        return await self._client.request(
            "GET",
            "/api/v1/usage/export",
            params={"format": format, "period": period},
        )

File: sdk/typescript/src/namespaces/usage.ts
/**
 * Usage Namespace API
 *
 * Provides methods for usage tracking and SME dashboard:
 * - Usage summary and breakdown
 * - ROI analysis
 * - Budget tracking
 * - Forecasting
 * - Industry benchmarks
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Get usage summary
 * const summary = await client.usage.getSummary({ period: 'month' });
 *
 * // Get ROI analysis
 * const roi = await client.usage.getROI();
 *
 * // Get budget status
 * const budget = await client.usage.getBudgetStatus();
 * ```
 */

/**
 * Interface for usage client methods.
 */
interface UsageClientInterface {
  get<T>(path: string, params?: Record<string, unknown>): Promise<T>;
  request<T>(method: string, path: string, options?: { params?: Record<string, unknown> }): Promise<T>;
}

/**
 * Time period options.
 */
export type UsagePeriod = 'day' | 'week' | 'month' | 'quarter' | 'year';

/**
 * Benchmark comparison types.
 */
export type BenchmarkType = 'sme' | 'enterprise' | 'tech_startup' | 'consulting';

/**
 * Export format options.
 */
export type ExportFormat = 'csv' | 'json';

/**
 * Group by dimension.
 */
export type GroupByDimension = 'agent' | 'model' | 'day' | 'debate_type';

/**
 * Usage summary response.
 */
export interface UsageSummary {
  period: string;
  total_debates: number;
  total_tokens: number;
  total_cost: number;
  active_users: number;
  avg_debate_duration_ms: number;
  consensus_rate: number;
  top_agents: Array<{ name: string; debates: number }>;
}

/**
 * Usage breakdown response.
 */
export interface UsageBreakdown {
  group_by: GroupByDimension;
  period: string;
  data: Array<{
    key: string;
    debates: number;
    tokens: number;
    cost: number;
  }>;
}

/**
 * ROI analysis response.
 */
export interface ROIAnalysis {
  time_saved_hours: number;
  cost_per_decision: number;
  benchmark_comparison: {
    industry_avg_cost: number;
    savings_percentage: number;
  };
  decision_quality_score: number;
  recommendations: string[];
}

/**
 * Budget status response.
 */
export interface BudgetStatus {
  daily_limit: number;
  daily_used: number;
  daily_remaining: number;
  monthly_limit: number;
  monthly_used: number;
  monthly_remaining: number;
  throttle_level: 'none' | 'light' | 'medium' | 'heavy' | 'blocked';
  reset_at: string;
}

/**
 * Usage forecast response.
 */
export interface UsageForecast {
  forecast_days: number;
  projected_debates: number;
  projected_cost: number;
  confidence: number;
  trend: 'increasing' | 'stable' | 'decreasing';
  daily_projections: Array<{
    date: string;
    projected_debates: number;
    projected_cost: number;
  }>;
}

/**
 * Industry benchmark data.
 */
export interface IndustryBenchmarks {
  benchmark_type: BenchmarkType;
  metrics: {
    avg_decisions_per_month: number;
    avg_cost_per_decision: number;
    avg_time_per_decision_hours: number;
    consensus_rate: number;
  };
  your_position: {
    percentile: number;
    comparison: 'above' | 'at' | 'below';
  };
}

/**
 * Export result.
 */
export interface UsageExport {
  format: ExportFormat;
  period: string;
  download_url: string;
  expires_at: string;
}

/**
 * Usage API for SME dashboard.
 *
 * Provides methods for tracking and analyzing usage:
 * - Summary metrics
 * - Breakdown by dimension
 * - ROI analysis
 * - Budget monitoring
 * - Forecasting
 */
export class UsageAPI {
  constructor(private client: UsageClientInterface) {}

  /**
   * Get unified usage metrics summary.
   *
   * @param options - Query options
   * @param options.period - Time period (day, week, month, quarter, year)
   * @param options.organizationId - Optional organization filter
   */
  async getSummary(options?: {
    period?: UsagePeriod;
    organizationId?: string;
  }): Promise<UsageSummary> {
    const params: Record<string, unknown> = {};
    if (options?.period) params.period = options.period;
    if (options?.organizationId) params.organization_id = options.organizationId;
    return this.client.request('GET', '/api/v1/usage/summary', { params });
  }

  /**
   * Get usage breakdown by dimension.
   *
   * @param options - Query options
   * @param options.groupBy - Dimension to group by
   * @param options.period - Time period
   */
  async getBreakdown(options?: {
    groupBy?: GroupByDimension;
    period?: UsagePeriod;
  }): Promise<UsageBreakdown> {
    const params: Record<string, unknown> = {};
    if (options?.groupBy) params.group_by = options.groupBy;
    if (options?.period) params.period = options.period;
    return this.client.request('GET', '/api/v1/usage/breakdown', { params });
  }

  /**
   * Get ROI analysis with industry benchmarks.
   *
   * @param options - Query options
   * @param options.benchmarkType - Benchmark to compare against
   */
  async getROI(options?: { benchmarkType?: BenchmarkType }): Promise<ROIAnalysis> {
    const params: Record<string, unknown> = {};
    if (options?.benchmarkType) params.benchmark_type = options.benchmarkType;
    return this.client.request('GET', '/api/v1/usage/roi', { params });
  }

  /**
   * Get current budget utilization status.
   */
  async getBudgetStatus(): Promise<BudgetStatus> {
    return this.client.get('/api/v1/usage/budget-status');
  }

  /**
   * Get usage forecast.
   *
   * @param options - Query options
   * @param options.days - Number of days to forecast (default: 30)
   */
  async getForecast(options?: { days?: number }): Promise<UsageForecast> {
    const params: Record<string, unknown> = {};
    if (options?.days) params.days = options.days;
    return this.client.request('GET', '/api/v1/usage/forecast', { params });
  }

  /**
   * Get industry benchmarks for comparison.
   *
   * @param options - Query options
   * @param options.benchmarkType - Industry benchmark type
   */
  async getBenchmarks(options?: { benchmarkType?: BenchmarkType }): Promise<IndustryBenchmarks> {
    const params: Record<string, unknown> = {};
    if (options?.benchmarkType) params.type = options.benchmarkType;
    return this.client.request('GET', '/api/v1/usage/benchmarks', { params });
  }

  /**
   * Export usage data.
   *
   * @param options - Export options
   * @param options.format - Export format (csv or json)
   * @param options.period - Time period to export
   */
  async export(options?: {
    format?: ExportFormat;
    period?: UsagePeriod;
  }): Promise<UsageExport> {
    const params: Record<string, unknown> = {};
    if (options?.format) params.format = options.format;
    if (options?.period) params.period = options.period;
    return this.client.request('GET', '/api/v1/usage/export', { params });
  }
}
